grouper/fe/handlers/user_metadata.py

```diff
@@ -0,0 +1,97 @@
+from __future__ import annotations
+
+from typing import TYPE_CHECKING
+
+from grouper.constants import USER_METADATA_GITHUB_USERNAME_KEY, USER_METADATA_SHELL_KEY
+from grouper.fe.forms import UserMetadataForm
+from grouper.fe.settings import settings
+from grouper.fe.util import GrouperHandler
+from grouper.models.audit_log import AuditLog
+from grouper.models.user import User
+from grouper.role_user import can_manage_role_user
+from grouper.service_account import can_manage_service_account
+from grouper.user_metadata import get_user_metadata_by_key, set_user_metadata
+
+if TYPE_CHECKING:
+    from grouper.models.base.session import Session
+    from typing import Any
+
+
+class UserMetadata(GrouperHandler):
+    @staticmethod
+    def check_access(session: Session, actor: User, target: User) -> bool:
+        return (
+            actor.name == target.name
+            or (target.role_user and can_manage_role_user(session, actor, tuser=target))
+            or (target.is_service_account and can_manage_service_account(session, target, actor))
+        )
+
+    def get(self, *args: Any, **kwargs: Any) -> None:
+        name = self.get_path_argument("name")
+
+        user = User.get(self.session, name=name)
+        if not user:
+            return self.notfound()
+
+        if not self.check_access(self.session, self.current_user, user):
+            return self.forbidden()
+
+        metadata_key = self.get_path_argument("key")
+
+        if metadata_key == USER_METADATA_SHELL_KEY:
+            return self.redirect("/users/{}/shell".format(user.name))
+        elif metadata_key == USER_METADATA_GITHUB_USERNAME_KEY:
+            return self.redirect("/github/link_begin/{}".format(user.id))
+
+        metadata_item = get_user_metadata_by_key(self.session, user.id, metadata_key)
+        if not metadata_item:
+            return self.notfound()
+
+        form = UserMetadataForm()
+        form.value.choices = settings().metadata_options[metadata_key]
+
+        self.render("user-metadata.html", form=form, user=user, metadata_key=metadata_key)
+
+    def post(self, *args: Any, **kwargs: Any) -> None:
+        name = self.get_path_argument("name")
+
+        user = User.get(self.session, name=name)
+        if not user:
+            return self.notfound()
+
+        if not self.check_access(self.session, self.current_user, user):
+            return self.forbidden()
+
+        metadata_key = self.get_path_argument("key")
+
+        if metadata_key == USER_METADATA_SHELL_KEY:
+            return self.redirect("/users/{}/shell".format(user.name))
+        elif metadata_key == USER_METADATA_GITHUB_USERNAME_KEY:
+            return self.redirect("/github/link_begin/{}".format(user.id))
+
+        metadata_item = get_user_metadata_by_key(self.session, user.id, metadata_key)
+        if not metadata_item:
+            return self.notfound()
```

Comment:
That's a good question, and something I was punting on, assuming we would always initialize fields we wanted to be settable. 

Two options I've considered:
1. The easy, possibly hacky way: add a new settings option which provides a list of default metadata fields that all users and/or service accounts should have.
2. Arguably the the "right" way (or "a" right way): extend the data model to support storing default metadata fields in the db.

Either of these seem like larger changes that would warrant more thought; -- it would be great if the model could extend to support the currently special cased shell (which should be simple enough), and github_username (less straightforward).

Reply:
Oh another thought -- add support in the UI for adding new metadata fields (something currently possible via `grouper-ctl`. Though this opens some questions around if we want arbitrary users/groups to be able to add new metadata fields, or if this should be controlled by the admins.

Reply:
Hmm. Letting arbitrary users create new metadata fields doesn't seem worth it unless we have a specific use case in mind; I think getting it to where changing the settings file is enough to create/allow a new metadata field would be fine.

I'm not sure (1) has to be hacky. How about a setting for "these metadata fields can be set regardless of whether the key currently exists"? Then you don't have to worry about setting the field on all existing users

Reply:
So what I ended up doing was have fields that are in the `metadata_options` settings show up in the UI as unset, with an option to set them. Does that sound ok for a compromise right now?